Implements dual-chunk-flash-attn backend for dual chunk attention with sparse attention support

sighingnow opened the pull request:
This PR implements the [dual-chunk flash attention](https://arxiv.org/pdf/2402.17463.pdf), a training-free method to extend model context length (see also #6139), with sparse attention (https://github.com/microsoft/MInference) support.

This PR requires the [sparse attention kernel](https://github.com/vllm-project/flash-attention/pull/33) from vllm-flash-attention. Qwen models with 1m context length support will be open-sourced in the next one or two weeks, and unit tests will be added later.

FIX #12452

jacob-crux commented:
I see that you have `enforce_eager=True` set, so it looks like there are still compatibility issues with cudagraph. 
Do you plan to fix this in the future?

sighingnow commented:
> I see that you have `enforce_eager=True` set, so it looks like there are still compatibility issues with cudagraph. Do you plan to fix this in the future?

All conflicts fixed, could you please take another look? thanks!

jacob-crux commented on vllm/attention/backends/dual_chunk_flash_attn.py:
```diff
@@ -0,0 +1,1439 @@
+                               st] = decode_metadata.block_tables[i, st:ed]
+        decode_metadata.block_tables_intra = block_tables_intra
+
+        seq_lens_succ = (chunk_num_curr -
```
When I try the Needle in a haystack test with qwen-7b and llama-8b(Modified code to support llama), there is a bug that produces a negative number when it is over 13k~15k.
I modified the code as below and confirmed that it works.

```
seq_lens_succ = ((chunk_num_curr - (chunk_num_curr - 1).clip(min=0)) * chunk_len)
```

jacob-crux commented:
> > I see that you have `enforce_eager=True` set, so it looks like there are still compatibility issues with cudagraph. Do you plan to fix this in the future?
> 
> All conflicts fixed, could you please take another look? thanks!

I tested it because I thought it was fixed, but I still have the same problem as below. 
Are you saying that Cudagraph capture is possible? (enforce_eager=False)

```
Capturing CUDA graph shapes:   0%|                                                                                                                                                                                                               | 0/35 [00:00<?, ?it/s]
[rank0]: Traceback (most recent call last):
[rank0]:   File "/data/lme-storage_810/jacob/needle/NeedleInAHaystack-lme/run_needle_in_haystack.py", line 435, in <module>
[rank0]:     ht = LLMNeedleHaystackTester(
[rank0]:          ^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/data/lme-storage_810/jacob/needle/NeedleInAHaystack-lme/run_needle_in_haystack.py", line 94, in __init__
[rank0]:     self.model_to_test = LLM(model=model_name)
[rank0]:                          ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/utils.py", line 1044, in inner
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/entrypoints/llm.py", line 228, in __init__
[rank0]:     self.llm_engine = self.engine_class.from_engine_args(
[rank0]:                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/engine/llm_engine.py", line 517, in from_engine_args
[rank0]:     engine = cls(
[rank0]:              ^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/engine/llm_engine.py", line 276, in __init__
[rank0]:     self._initialize_kv_caches()
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/engine/llm_engine.py", line 429, in _initialize_kv_caches
[rank0]:     self.model_executor.initialize_cache(num_gpu_blocks, num_cpu_blocks)
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/executor/gpu_executor.py", line 83, in initialize_cache
[rank0]:     self.driver_worker.initialize_cache(num_gpu_blocks, num_cpu_blocks)
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/worker/worker.py", line 274, in initialize_cache
[rank0]:     self._warm_up_model()
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/worker/worker.py", line 292, in _warm_up_model
[rank0]:     self.model_runner.capture_model(self.gpu_cache)
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/worker/model_runner.py", line 1533, in capture_model
[rank0]:     graph_runner.capture(**capture_inputs)
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/worker/model_runner.py", line 1885, in capture
[rank0]:     self.model(
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1747, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/model_executor/models/qwen2.py", line 496, in forward
[rank0]:     hidden_states = self.model(input_ids, positions, kv_caches,
[rank0]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/compilation/decorators.py", line 170, in __call__
[rank0]:     return self.forward(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/model_executor/models/qwen2.py", line 359, in forward
[rank0]:     hidden_states, residual = layer(
[rank0]:                               ^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1747, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/model_executor/models/qwen2.py", line 267, in forward
[rank0]:     hidden_states = self.self_attn(
[rank0]:                     ^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1747, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/model_executor/models/qwen2.py", line 189, in forward
[rank0]:     attn_output = self.attn(q,
[rank0]:                   ^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1747, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/attention/layer.py", line 185, in forward
[rank0]:     return torch.ops.vllm.unified_attention(
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/usr/local/lib/python3.12/dist-packages/torch/_ops.py", line 1116, in __call__
[rank0]:     return self._op(*args, **(kwargs or {}))
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/attention/layer.py", line 280, in unified_attention
[rank0]:     return self.impl.forward(query, key, value, kv_cache, attn_metadata,
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/bc-user/vllm_dual_chunk_250114/vllm/vllm/attention/backends/dual_chunk_flash_attn.py", line 373, in forward
[rank0]:     assert decode_meta.scaling_factor is not None
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]: AssertionError
```

sighingnow commented:
Rebase against main.

Hi @youkaichao @simon-mo @WoosukKwon Do you folks think if there are still things that need to be improved in this pull reqeust?

Thanks!

tlrmchlsmth commented on csrc/attention/vertical_slash_index.cu:
```diff
@@ -0,0 +1,394 @@
+
+  bool slash_finished = false;
+  while (1) {
+    // if (v > 100) return;
```
ditto - debug cruft

tlrmchlsmth commented on csrc/attention/vertical_slash_index.cu:
```diff
@@ -0,0 +1,394 @@
+    int* column_index,  // [BATCH, N_HEADS, cdiv(N_CTX, BLOCK_SIZE_M), NNZ_V]
+    int64_t BATCH_SIZE, int64_t N_HEADS, int64_t N_ROWS, int64_t BLOCK_SIZE_M,
+    int64_t BLOCK_SIZE_N, int64_t NNZ_V, int64_t NNZ_S, bool causal) {
+  // const int BLOCK_SIZE_M = 64;
+  // const int BLOCK_SIZE_N = 64;
```
ditto

tlrmchlsmth commented on csrc/attention/vertical_slash_index.cu:
```diff
@@ -0,0 +1,394 @@
+    torch::Tensor slash_indexes,     // [BATCH, N_HEADS, NNZ_S]
+    int64_t context_size, int64_t block_size_M, int64_t block_size_N,
+    bool causal) {
+  // assert(block_size_M == 64);
+  // assert(block_size_N == 64);
```
ditto

tlrmchlsmth commented on csrc/attention/vertical_slash_index.cu:
```diff
@@ -0,0 +1,394 @@
+    torch::Tensor vertical_indices_count,  // [N_HEADS, ]
+    torch::Tensor slash_indices_count, int64_t context_size,
+    int64_t block_size_M, int64_t block_size_N, bool causal) {
+  // assert(block_size_M == 64);
+  // assert(block_size_N == 64);
```
ditto

tlrmchlsmth reviewed (commented):
Spotted a few bits ofcommented out code that look like debug cruft or are otherwise mysterious. Could you clean those up and any other similar spots?

tlrmchlsmth commented on examples/offline_inference_qwen_1m.py:
```diff
@@ -0,0 +1,36 @@
+os.environ["VLLM_ATTENTION_BACKEND"] = "DUAL_CHUNK_FLASH_ATTN"
+os.environ["VLLM_ALLOW_LONG_MAX_MODEL_LEN"] = "1"
+
+with open(os.path.expanduser("~/vllm/64k.txt")) as f:
```
This should be updated to a proper URL

tlrmchlsmth commented on examples/offline_inference_qwen_1m.py:
```diff
@@ -0,0 +1,36 @@
+)
+
+# Create an LLM.
+llm = LLM(model=os.path.expanduser("~/models/qwen2.5-14b-1m-1231/"),
```
ditto

tlrmchlsmth commented on examples/offline_inference_qwen_1m.py:
I think this should be moved to `examples/offline_inference/qwen_1m.py`

tlrmchlsmth commented on vllm/attention/backends/flash_attn.py:
```diff
@@ -105,6 +105,11 @@ class FlashAttentionMetadata(AttentionMetadata):
     # seq_lens stored as a tensor.
     seq_lens_tensor: Optional[torch.Tensor]
 
+    # (batch_size,). The orig sequence length per sequence.
+    orig_seq_lens: Optional[List[int]]
+    # orig_seq_lens stored as a tensor.
+    orig_seq_lens_tensor: Optional[torch.Tensor]
```
IMO this PR shouldn't need to touch the `flash_attn` backend. Could you say a bit about why you added these?

sighingnow replied:
I added the orig_seq_lens to the flash-attn backend metadata then dual-chunk-flash-attn can inherit most of the logic of metadata building from flash-attn to avoid code duplication.

tlrmchlsmth commented on vllm/engine/arg_utils.py:
```diff
@@ -1035,6 +1036,12 @@ def create_engine_config(self,
                 "BitsAndBytes load format and QLoRA adapter only support "
                 f"'bitsandbytes' quantization, but got {self.quantization}")
 
+        if envs.VLLM_ATTENTION_BACKEND in [STR_DUAL_CHUNK_FLASH_ATTN_VAL]:
+            assert self.enforce_eager, (
+                "Cuda graph is not supported with DualChunkFlashAttention. "
+                "To run the model in eager mode, set 'enforce_eager=True' "
+                "or use '--enforce-eager' in the CLI.")
```
Do you know what the blockers are for Cuda graphs + DualChunkFlashAttention?

sighingnow replied:
The current implementation doesn't support cuda graph for two reasons: (1) cuda graph is designed for short sequence length (due to block tables padding), the default capture sequence length is 8192, and, (2) there are some CPU logic (about sequence length) inside the DualChunkFlashAttention backend.

LucasWilkinson commented on vllm/worker/model_runner.py:
```diff
@@ -491,6 +491,7 @@ def _compute_lens(self, inter_data: InterDataForSeqGroup, seq_idx: int,
         # already computed) and sequence length (total number of tokens).
 
         seq_len = seq_data.get_len()
+        orig_seq_len = seq_data.get_len()
```
I personally find the `orig_seq_lens` vs `seq_lens` naming confusing. Given that this touches core code such as `model_runner.py` I think we should be very thoughtful in our naming.

in my opinion I would  get rid of `orig_seq_lens` in `InterDataForSeqGroup` and add `prompt_lens` instead, that would be:

```
prompt_lens = seq_data.get_prompt_len()
```

then move 
```
    # (batch_size,). The orig sequence length per sequence.
    orig_seq_lens: Optional[List[int]]
    # orig_seq_lens stored as a tensor.
    orig_seq_lens_tensor: Optional[torch.Tensor]
```

into the `DualChunkFlashAttentionMetadata`, I think the `DualChunkFlashAttentionMetadataBuilder` should be able to reconstruct `orig_seq_lens` by doing something like (untested):
```
class DualChunkFlashAttentionMetadataBuilder(FlashAttentionMetadataBuilder):
    def __init__(self, input_builder: "ModelInputForGPUBuilder"):
      super().__init__(input_builder)
      self.orig_seq_lens: List[int] = []

    def _add_seq_group(
            self, inter_data: "ModelInputForGPUBuilder.InterDataForSeqGroup",
            chunked_prefill_enabled: bool, prefix_cache_hit: bool):
      super()._add_seq_group(inter_data, chunked_prefill_enabled, prefix_cache_hit)
      for prompt_len, seq_len in zip(inter_data.prompt_lens, inter_data.seq_lens):
        self.orig_seq_lens.append(max(prompt_len, seq_len))

    def build(...): ... # update as appropriate

```

This would atleast isolate `orig_seq_lens` to the dual chunk code (i would consider renaming the this to `full_seq_lens` but thats personal-preference/nit). 

By using `prompt_lens` in `InterDataForSeqGroup` atleast we are not introducing new terminology to the `model_runner.py` code since `seq_data.get_prompt_len()` already exists as a concept.

And by moving `orig_seq_lens`/`orig_seq_lens_tensor` down into `DualChunkFlashAttentionMetadataBuilder` and `DualChunkFlashAttentionMetadata` we can avoid touching `flash_attn.py` and `xformers.py` 

@tlrmchlsmth do you have an opinion on this?

tlrmchlsmth replied:
I think this sounds like a great approach if the authors agree! Really appreciate the concrete suggestion

sighingnow replied:
I would take a try to do such refactor, thanks for the suggestion.

LucasWilkinson commented on vllm/attention/layer.py:
```diff
@@ -95,15 +96,20 @@ def __init__(
+        self.impl = impl_cls(
+            num_heads, head_size, scale, num_kv_heads, alibi_slopes,
+            sliding_window, kv_cache_dtype, blocksparse_params,
+            logits_soft_cap, attn_type, **{
+                "dual_chunk_attention_config": dual_chunk_attention_config,
+                "prefix": prefix,
+            } if dual_chunk_attention_config is not None else {})
```
I feel like this messy, I think we should maybe do something like:

```
def __init__(..., **extra_attn_kwargs):
   self.impl = impl_cls(..., **extra_attn_kwargs)
```
the challenge here is `prefix` would not be captured by `extra_attn_kwargs` but is only (currently) used by `DualChunkFlashAttentionImpl`. I do think it would be less messy though to do this any make `prefix` a standard arg for attention impls, given that it is pretty generic. Thoughts @WoosukKwon

LucasWilkinson commented on vllm/attention/layer.py:
```diff
@@ -134,7 +140,22 @@ def forward(
+                                             torch.Tensor]] = None,
     ) -> torch.Tensor:
+
+        if self.dual_chunk_attention_config:
+            assert query_succ_and_inter is not None
+            dca_kwargs = {
+                "query_succ": query_succ_and_inter[0],
+                "query_inter": query_succ_and_inter[1],
+                "query_succ_critical": query_succ_and_inter[2],
+                "query_inter_critical": query_succ_and_inter[3],
+            } if query_succ_and_inter else {}
+        else:
+            dca_kwargs = {}
+
```
I think we should try hard to see if there is cleaner way of passing these, maybe they can be bundled into a single `q` tensor that get reinterpreted as components via a combination of slicing and `.view` calls in the attn impl?

sighingnow replied:
I would take a try to see if it can be simplified.

LucasWilkinson commented on CMakeLists.txt:
```diff
@@ -550,7 +551,7 @@ else()
           vllm-flash-attn
           GIT_REPOSITORY https://github.com/vllm-project/flash-attention.git
-          GIT_TAG 96266b1111111f3d11aabefaf3bacbab6a89d03c
+          GIT_TAG 323b789ae92b0376c0adfcfe1b48571ac32dc411
```
with the landing of FA3 Support (https://github.com/vllm-project/vllm/pull/12093) the vllm-flash-attn repo is going to be "reset", the `lwilkinson/fa3-squashed` will become the new main branch (via renaming)

I cherry picked the sparse attention kernels over to that branch so the commit used by current vllm main `90eacc1af2a7c3de62ea249e929ed5faccf38954` should support what you need, can you please rebase and swap to using this commit?  please let me know if this causes any issues incase the cherry pick broke something (the unit tests did pass)

sighingnow replied:
Thank you, will do it.

sighingnow committed:
Implements dual-chunk-flash-attn backend for dual chunk attention with sparse attention support.

Signed-off-by: Tao He <linzhu.ht@alibaba-inc.com>

sighingnow commented:
Hi @LucasWilkinson most of the comments has been addressed, could you please take another look? Thanks!

The lint error comes from the prompt text, do you have any suggestion about how could I skip/resolve it?

ywang96 commented:
@sighingnow Sorry for the delayed response! I've merged main into your branch so the pre-commit should be cleared. I'll enable ready status for this PR so at least we can get the CI going before @tlrmchlsmth or @LucasWilkinson want to give their final greenlight!

tlrmchlsmth commented on examples/offline_inference/qwen_1m/1m.txt:
I think we should revisit whether it makes sense to have this 1m context example here. This file would be by far the largest in the vLLM git repo at 4MB, and I think it might be better for the example to pull the data from an external source.

Side note, I thought having the example prompts start by discussing the differences between various narcotics was an interesting choice!

tlrmchlsmth commented on vllm/_custom_ops.py:
```diff
@@ -122,6 +122,101 @@ def paged_attention_rocm(
+    kv_seqlens: torch.Tensor,  # [BATCH, ]
+    vertical_indexes: torch.Tensor,  # [BATCH, N_HEADS, NNZ_V]
+    slash_indexes: torch.Tensor,  # [BATCH, N_HEADS, NNZ_S]
+    vertical_indices_count: torch.
+    Tensor,  # [N_HEADS] : different head use different number of indices
```
nit: some weird whitespace here - I wouldn't want us to spend too much effort fighting clang-format but maybe the comment can be moved to its own line

tlrmchlsmth commented on vllm/attention/backends/dual_chunk_flash_attn.py:
```diff
@@ -0,0 +1,1484 @@
+                dtype=decode_metadata.block_tables.dtype,
+                device=decode_metadata.block_tables.device,
+            )
+            for i in range(batch_size):
+                st = ((chunk_num_curr[i] - 1).clip(min=0) * chunk_len //
+                      self.block_size)
+                ed = min(
+                    st + (max_seq_len_succ - 1) // self.block_size + 1,
+                    (cache_seq_lens[i] - 1) // self.block_size + 1,
+                )
+                block_tables_succ[i, :ed -
+                                  st] = decode_metadata.block_tables[i, st:ed]
```
Improved readability:
```suggestion
            for i in range(batch_size):
                start = ((chunk_num_curr[i] - 1).clip(min=0) * chunk_len //
                      self.block_size)
                end = min(
                    start + (max_seq_len_succ - 1) // self.block_size + 1,
                    (cache_seq_lens[i] - 1) // self.block_size + 1,
                )
                block_tables_succ[i, :end -
                                  start] = decode_metadata.block_tables[i, start:end]
```

tlrmchlsmth commented on vllm/attention/backends/dual_chunk_flash_attn.py:
```diff
@@ -0,0 +1,1484 @@
+                succ_vertical_indices, succ_slash_indices = None, None
+                inter_vertical_indices, inter_slash_indices = None, None
+
+            if sparse_attn_enabled:
+                flash_result = self._do_flash_attn(
+                    q_states_intra,
+                    k_states_intra,
+                    v_states_intra,
+                    softmax_scale=softmax_scale,
+                    causal=True,
+                    stage="intra",
+                    vertical_indices=vertical_buffer,
+                    slash_indices=slash_buffer,
+                    vertical_indices_count=vertical_size_buffer,
+                    slash_indices_count=slash_sizes_buffer,
+                    mergehead_softmax_scale=softmax_scale,
+                    sparse_attn_enabled=sparse_attn_enabled)
+            else:
+                flash_result = self._do_flash_attn(
+                    q_states_intra,
+                    k_states_intra,
+                    v_states_intra,
+                    softmax_scale=softmax_scale,
+                    causal=True,
+                    stage="intra",
+                    vertical_indices=intra_vertical_indices,
+                    slash_indices=intra_slash_indices,
+                    sparse_attn_enabled=sparse_attn_enabled)
```
QQ: is the if/else necessary since `sparse_attn_enabled` is passed into the function in both cases?

tlrmchlsmth commented on vllm/attention/backends/dual_chunk_flash_attn.py:
Please add SPDX headers to new .py files (see https://github.com/vllm-project/vllm/pull/12628 for reference)

tlrmchlsmth reviewed (commented):
A few more review comments, mostly minor stuff. Looks pretty good, although I do suggest getting rid of the `qwen_1m` example

mklasby commented on vllm/attention/backends/dual_chunk_flash_attn.py:
```diff
@@ -0,0 +1,1484 @@
+                        softmax_scale,
+                        chunk_size,
+                        local_size,
+                        scaling_factor.item(),
```
```suggestion
                        scaling_factor[i].item(),
```

Otherwise this will error since scaling_factor contains more than 1 element here

freedomkk-qfeng commented:
Any progress?

Ki6an commented:
+1

LucasWilkinson commented on vllm/config.py:
```diff
@@ -739,6 +739,23 @@ def _verify_with_expert_parallelism(self) -> None:
                 "Number of experts in the model must be greater than 0 "
                 "when expert parallelism is enabled.")
 
+    def verify_dual_chunk_attention_config(
```
nit: I dont love this but not really sure if theres a better place (any ideas @mgoin ?)

LucasWilkinson reviewed (commented):
I think this is getting very close, thanks for rebasing it! My main concern right now is the large text files in the repo. Also there appear to still be unaddressed review comments from before, please ping us when this is ready for final review.

sighingnow commented:
> I think this is getting very close, thanks for rebasing it! My main concern right now is the large text files in the repo. Also there appear to still be unaddressed review comments from before, please ping us when this is ready for final review.

Hi @LucasWilkinson, thanks for these comments. I have rebased this branch over current main, removed those example prompts and provided them as URLs, and address the reviewer comments above in this PR. Now I think it should be ready for landing.

Before landing, a bugfix in flash-attention should be merged first: https://github.com/vllm-project/flash-attention/pull/60. After that, I will revise the dependency version of vllm-flash-attention in this PR.

tlrmchlsmth reviewed (commented):
Thank you for rebasing on current main! The code looks pretty clean to me now.

Before this lands, I think we should make sure there's a plan in place to get support for this in vLLM V1. vLLM has switched to V1 by default and we are trying to deprecate V0.

A couple of questions:
* What will happen with this PR when running Qwen2 on systems where the dual-chunk attention backend is not supported? (e.g. AMD GPUs, TPUs, etc)
* Does vLLM automatically fall back to V0 when using dual-chunk attention?

sighingnow commented on examples/offline_inference/qwen_1m.py:
```diff
@@ -0,0 +1,65 @@
+    # https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2.5-1M/test-data/600k.txt
+    # https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2.5-1M/test-data/1m.txt
+
+    with urlopen("https://qianwen-res.oss-cn-beijing.aliyuncs.com"
+                 "/Qwen2.5-1M/test-data/600k.txt") as response:
```
Fixed.

sighingnow commented:
> A couple of questions:
> 
> * What will happen with this PR when running Qwen2 on systems where the dual-chunk attention backend is not supported? (e.g. AMD GPUs, TPUs, etc)
> * Does vLLM automatically fall back to V0 when using dual-chunk attention?

We have launched the work of migrating qwen related changes in our internal repo to v1 since v1 becomes the default option in vLLM. The dual-chunk-attn backend would be adapted to v1, too, and most of changesets could be reused.

I have added an assertion in arg_utils.py to check if the current platform is cuda(the sparse_attn_func is only available in vllm-project/flash-attention for cuda) and if the current engine is v0.

LucasWilkinson commented:
https://github.com/vllm-project/flash-attention/pull/60 has landed can you please update this PR?

sighingnow commented:
> [vllm-project/flash-attention#60](https://github.com/vllm-project/flash-attention/pull/60) has landed can you please update this PR?

Done, and rebased to main.

sighingnow commented:
@LucasWilkinson I have rebased to current main again. Could you please take another look on this PR? Thanks!

LucasWilkinson commented on cmake/external_projects/vllm_flash_attn.cmake:
This can actually be left as `0a721daebe4fa7149f06ecf3d3eabeb6dcd0f1fa` since that includes the PR you need

LucasWilkinson reviewed (approved):
Apologies overall this looks good now! Thanks for all the updates, the only things left to see on my end would be:

- [ ] Address: https://github.com/vllm-project/vllm/pull/11844/files#r2070914950
- [ ] Rebase
- [ ] Address: https://github.com/vllm-project/vllm/pull/11844/files#r1943492013
- [ ] @mgoin address: https://github.com/vllm-project/vllm/pull/11844/files#r2039650211

sighingnow commented:
> @sighingnow Thanks for the update! looking into the CI failure it does not appear to be related (V1 code, this PR does not touch V1) but this is a bit out of my area of expertise, asking around (cc @russellb)

Rebased against main again. The failed test cases shouldn't be caused by this PR. It failed on a speculative decoding cases and seems that that case is not executed by all PRs.

ExtReMLapin commented:
So If I understand correctly, now Qwen2.5-1M actually uses the correct attention mechanism and VRAM should be lowered and prompt processing faster, right ?

exceedzhang commented:
I tested Qwen/Qwen2.5-7B-Instruct-1M using DualChunkFlashAttention backend.
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/a439f5f0-78e1-4606-95d8-abbbddf4bc45" />
It startup well, but not work well. @sighingnow 

ubuntu-vllm-openai-1  | INFO 05-31 19:13:07 [logger.py:42] Received request cmpl-77d91882816c4f748e2023c93449f62d-0: prompt: 'Once upon a time', params: SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.05, temperature=0.0, top_p=1.0, top_k=0, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=1000, min_tokens=0, logprobs=1, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None, extra_args=None), prompt_token_ids: [12522, 5193, 264, 882], prompt_embeds shape: None, lora_request: None, prompt_adapter_request: None.
ubuntu-vllm-openai-1  | INFO 05-31 19:13:07 [engine.py:316] Added request cmpl-77d91882816c4f748e2023c93449f62d-0.
ubuntu-vllm-openai-1  | INFO:     172.18.0.1:46884 - "POST /v1/completions HTTP/1.1" 500 Internal Server Error
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164] AssertionError('seqused_k must be provided if block_table is provided')
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164] Traceback (most recent call last):
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 162, in start
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     self.run_engine_loop()
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 225, in run_engine_loop
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     request_outputs = self.engine_step()
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                       ^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 251, in engine_step
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     raise e
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 234, in engine_step
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self.engine.step()
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/llm_engine.py", line 1393, in step
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     outputs = self.model_executor.execute_model(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/executor_base.py", line 299, in execute_model
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     driver_outputs = self._driver_execute_model(execute_model_req)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/mp_distributed_executor.py", line 144, in _driver_execute_model
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self.driver_worker.execute_model(execute_model_req)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker_base.py", line 420, in execute_model
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     output = self.model_runner.execute_model(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return func(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/model_runner.py", line 1843, in execute_model
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     hidden_or_intermediate_states = model_executable(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                                     ^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self._call_impl(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1762, in _call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return forward_call(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 481, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     hidden_states = self.model(input_ids, positions, intermediate_tensors,
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/decorators.py", line 172, in __call__
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self.forward(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 358, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     hidden_states, residual = layer(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                               ^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self._call_impl(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1762, in _call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return forward_call(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 257, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     hidden_states = self.self_attn(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                     ^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self._call_impl(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1762, in _call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return forward_call(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 187, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     attn_output = self.attn(q, k, v)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                   ^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self._call_impl(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1762, in _call_impl
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return forward_call(*args, **kwargs)
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/layer.py", line 237, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return torch.ops.vllm.unified_attention(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/torch/_ops.py", line 1158, in __call__
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     return self._op(*args, **(kwargs or {}))
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/layer.py", line 386, in unified_attention
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     output = self.impl.forward(self, query, key, value, kv_cache,
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/dual_chunk_flash_attn.py", line 493, in forward
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     self._dual_chunk_flash_attn_prefill(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/dual_chunk_flash_attn.py", line 673, in _dual_chunk_flash_attn_prefill
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     current_out = self._dual_chunk_flash_attn_prefill_func(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/dual_chunk_flash_attn.py", line 1055, in _dual_chunk_flash_attn_prefill_func
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     flash_result = self._do_flash_attn(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                    ^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/dual_chunk_flash_attn.py", line 1207, in _do_flash_attn
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     output, softmax_lse = flash_attn_varlen_func(
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]                           ^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]   File "/usr/local/lib/python3.12/dist-packages/vllm/vllm_flash_attn/flash_attn_interface.py", line 204, in flash_attn_varlen_func
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]     assert block_table is None or seqused_k is not None, \
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ubuntu-vllm-openai-1  | ERROR 05-31 19:13:08 [engine.py:164] AssertionError: seqused_k must be provided if block_table is provided

ExtReMLapin commented:
Not a "blog" but it can help people working with it, so far we got much better results with Qwen 2.5 7b 1m than with nemotron 4M from nvidia.

However beside the issues states before with quantization and gpu splitting, we did not manage either to do batching/parallel processing

sighingnow commented:
Thanks for reporting @ExtReMLapin @exceedzhang . Will investigate this week.

ExtReMLapin commented:
@sighingnow @exceedzhang  thank for your contributions, it's mostly those PR that needs a review : 

Priority because crash fix : https://github.com/vllm-project/vllm/pull/19084

FP8 quantization support https://github.com/vllm-project/vllm/pull/19420

ExtReMLapin commented:
Got it, not merging this performance branch into the FP8 branch then, it's not worth the risk of breaking something !

Again at the office we really appreciate the effort spent on releasing those models.

We can a lot of tests, including other models claiming to have long context :

* Llama-3.1-Nemotron-8B-UltraLong-1M-Instruct : Not following instructions correctly
* gradientai/Llama-3-8B-Instruct-262k following instructions but struggles to speak anything else than english
* 01-ai/Yi-9B-200K byebye template chat being broken
* phi-3 128k not enough vram for 128k context
* Menlo/Jan-nano-128k really meh result, not following instructions correctly
* aws-prototyping/MegaBeam-Mistral-7B-512k same issues as above


 but this is the **ONLY** model actually following instructions on very long context and that can be ran easily (not insane resources).

Looking forward to see more models like this in the future !

ExtReMLapin commented:
Support added to latest 30B-A3B
https://huggingface.co/Qwen/Qwen3-30B-A3B-Instruct-2507/commit/3ffd1f50b179e643d839c86df9ffbbefcb0d5018

ExtReMLapin commented:
Any news on V1 implementation ?

Xuweijia-buaa commented on csrc/attention/vertical_slash_index.cu:
This function turns slash lines and vertical lines into real KV ranges to prepare inputs for sparse attention.

Xuweijia-buaa replied:
We plan to refine this when implementing sparse attention for vLLM v1.
